File: packages/flutter/lib/src/internal/keys.dart
/// The shortcut vocabulary, read rather than written.
///
/// [PlHotKeys] turns `Mod+Enter` into a pair of key caps; this turns the same
/// string into a predicate over a real key event. They share this file because a
/// shortcut a widget **displays** and a shortcut it **binds** have to be spelled
/// the same way — otherwise the cap on the screen is a claim nobody checked.
/// `PlCommandPalette` made that argument first and carried a private copy of the
/// matcher to make it; the fields that take a `hotKeys` map are the reason it
/// lives here instead.
///
/// Nothing here is exported from `plass_ui.dart` except [PlassHotKeys], which is
/// the map a caller writes, and which `types.dart` names.
library;

import 'package:flutter/foundation.dart';
import 'package:flutter/services.dart';
import 'package:flutter/widgets.dart';

import 'package:plass_ui/src/types.dart';

/// Whether this host spells `Mod` as ⌘.
///
/// Read from [defaultTargetPlatform], which is also where `PlHotKeys.platform`
/// reads it, so a `debugDefaultTargetPlatformOverride` in a test or a preview
/// moves the cap and the binding together rather than one of them.
bool get plassModIsMeta {
  return defaultTargetPlatform == TargetPlatform.macOS ||
      defaultTargetPlatform == TargetPlatform.iOS;
}

/// The spellings that mean the same key, folded onto one name.
///
/// A caller writes what is printed on their own keyboard — `Cmd`, `Option`,
/// `Return`, `Esc` — and every one of those has to reach the same cap and the
/// same binding.
const Map<String, String> plassKeyAliases = <String, String>{
  'cmdorctrl': 'mod',
  'commandorcontrol': 'mod',
  'cmd': 'meta',
  'command': 'meta',
  'super': 'meta',
  'win': 'meta',
  'windows': 'meta',
  'control': 'ctrl',
  'option': 'alt',
  'opt': 'alt',
  'return': 'enter',
  // A numpad Enter is an Enter. The framework labels it apart; a key cap does
  // not, and neither does anyone reaching for it.
  'numpadenter': 'enter',
  'esc': 'escape',
  'del': 'delete',
  'caps': 'capslock',
};

/// One token of a chord, folded: case, spacing and the aliases above.
String plassCanonicalKey(String token) {
  final normalized = token.toLowerCase().replaceAll(RegExp(r'[\s_-]'), '');

  return plassKeyAliases[normalized] ?? normalized;
}

/// The handful of keys whose `keyLabel` is not what is printed on the cap.
///
/// A cap says `↑` and the framework says `Arrow Up`, which folds to `arrowup`;
/// a cap says `Space` and the framework says a literal space. Everything else —
/// `Enter`, `Escape`, `Tab`, a letter — matches once both sides are folded.
const Map<String, String> _labelKeys = <String, String>{
  'up': 'arrowup',
  'down': 'arrowdown',
  'left': 'arrowleft',
  'right': 'arrowright',
  'space': ' ',
  'plus': '+',
};

/// Is this the chord?
///
/// Every modifier is checked in **both** directions, which is the difference
/// between binding a shortcut and binding a key: `'Enter'` must not fire on
/// `Mod+Enter`, or a field that saves on Enter also saves on every chord that
/// happens to end in one.
bool plassMatchesHotKey(String chord, KeyEvent event) {
  final List<String> parts = chord.split('+').map(plassCanonicalKey).toList(growable: false);
  final Set<String> wanted = parts.sublist(0, parts.length - 1).toSet();
  final String last = parts.last;
  final String key = _labelKeys[last] ?? last;

  final HardwareKeyboard keyboard = HardwareKeyboard.instance;
  final bool mod = plassModIsMeta ? keyboard.isMetaPressed : keyboard.isControlPressed;

  if (wanted.contains('mod') != mod) {
    return false;
  }

  if (wanted.contains('shift') != keyboard.isShiftPressed ||
      wanted.contains('alt') != keyboard.isAltPressed) {
    return false;
  }

  // Only when `Mod` was not asked for. It *is* one of these two, and checking
  // both would make `Mod+K` require a Ctrl that a Mac never sends.
  if (!wanted.contains('mod')) {
    if (wanted.contains('ctrl') != keyboard.isControlPressed ||
        wanted.contains('meta') != keyboard.isMetaPressed) {
      return false;
    }
  }

  return _foldLabel(event.logicalKey.keyLabel) == key;
}

/// What the framework calls a key, folded onto the name a cap is written with.
///
/// A one-character label is lower-cased and left alone — anything else would
/// turn the space bar's own label, which is a space, into an empty string. The
/// rest goes through the full fold, so `Arrow Up` reaches `arrowup` and
/// `Numpad Enter` reaches `enter` by way of the alias table.
String _foldLabel(String label) {
  return label.length == 1 ? label.toLowerCase() : plassCanonicalKey(label);
}

/// Does [hotKeys] claim [key] with no modifiers on it?
///
/// A control that binds its own activation keys through a `Shortcuts` asks this
/// so it can **stand down**. Of two handlers the one nearer the focused node
/// answers first, and a `FocusableActionDetector`'s is nearer than anything a
/// field can wrap around it — so the only way a caller's bare `Enter` can be
/// theirs is for the control to stop claiming it while they are asking for it.
///
/// [key] is a canonical name, so `'enter'` and never `'Enter'`.
bool plassClaimsBareKey(PlassHotKeys? hotKeys, String key) {
  if (hotKeys == null) {
    return false;
  }

  return hotKeys.keys.any((String chord) {
    final List<String> parts = chord.split('+');

    return parts.length == 1 && plassCanonicalKey(parts.single) == key;
  });
}

/// Runs whichever chord this key event is, and says whether one was.
///
/// [KeyEventResult.handled] on a match, which is what consuming it means: the
/// key reaches neither the control's own key handling nor the route above it, so
/// `Escape` bound on a field does not also close the dialog around it.
///
/// Repeats are answered as well as presses — holding a chord down that steps a
/// value should step it — and key *ups* never are.
KeyEventResult plassHandleHotKeys(PlassHotKeys? hotKeys, KeyEvent event) {
  if (hotKeys == null || hotKeys.isEmpty) {
    return KeyEventResult.ignored;
  }

  if (event is! KeyDownEvent && event is! KeyRepeatEvent) {
    return KeyEventResult.ignored;
  }

  for (final MapEntry<String, VoidCallback> entry in hotKeys.entries) {
    if (plassMatchesHotKey(entry.key, event)) {
      entry.value();

      return KeyEventResult.handled;
    }
  }

  return KeyEventResult.ignored;
}

/// The [Focus] a control carrying a `hotKeys` map wears.
///
/// Wrapped **inside** the field rather than around it, for the reason
/// `PlNumberField`'s `Shortcuts` is: of two handlers, the one nearer the focused
/// node answers first — so a chord bound here beats the app's own text-editing
/// shortcuts and the route's `Escape`, which is what consuming a key means.
///
/// It never takes the focus itself and never appears in the traversal order.
/// The control below it is the thing a reader is typing into, and a second stop
/// in front of it would be a Tab that goes nowhere.
///
/// Returns the child untouched when there is nothing to bind, so a field with no
/// `hotKeys` gains no node at all.
Widget plassHotKeyScope({required PlassHotKeys? hotKeys, required Widget child}) {
  if (hotKeys == null || hotKeys.isEmpty) {
    return child;
  }

  return Focus(
    canRequestFocus: false,
    skipTraversal: true,
    onKeyEvent: (FocusNode node, KeyEvent event) => plassHandleHotKeys(hotKeys, event),
    child: child,
  );
}

File: packages/react/src/internal/keys.ts
/**
 * The shortcut vocabulary, read rather than written.
 *
 * `PlHotKeys` turns `Mod+Enter` into a pair of key caps; this turns the same
 * string into a predicate over a real keyboard event. They are one file because
 * a shortcut a component **displays** and a shortcut it **binds** have to be
 * spelled the same way — otherwise the cap on the screen is a claim nobody
 * checked. `PlCommandPalette` made that argument first and carried a private
 * copy of the matcher to make it; the fields that take a `hotKeys` map are the
 * reason it lives here instead, and `PlHotKeys` now reads the alias table and
 * the platform out of this file rather than owning them.
 *
 * Nothing here is exported from `src/index.ts`. The one public name is
 * `PlassHotKeys` in `src/types.ts`, which is the map a caller writes.
 */

import type * as React from 'react';
import type { PlassHotKeys } from '../types.js';

/** The three real platforms, once `auto` has been resolved. */
export type PlassOS = 'mac' | 'windows' | 'linux';

/**
 * What the browser says it is running on.
 *
 * `userAgentData.platform` is the modern spelling and `navigator.platform` the
 * deprecated one that every browser still answers; the user agent string is the
 * last resort. All three are matched at once because the question here is coarse
 * — which of three key caps to print, and which modifier `Mod` stands for — and
 * getting it slightly wrong is a label, not a bug.
 */
export function detectOS(): PlassOS {
  if (typeof navigator === 'undefined') {
    return 'windows';
  }

  const data = (navigator as Navigator & { userAgentData?: { platform?: string } }).userAgentData;
  const haystack = `${data?.platform ?? ''} ${navigator.platform ?? ''} ${navigator.userAgent ?? ''}`;

  if (/mac|iphone|ipad|ipod/i.test(haystack)) {
    return 'mac';
  }

  if (/win/i.test(haystack)) {
    return 'windows';
  }

  return 'linux';
}

/**
 * The spellings that mean the same key, folded onto one name.
 *
 * A caller writes what is printed on their own keyboard — `Cmd`, `Option`,
 * `Return`, `Esc` — and every one of those has to reach the same cap and the
 * same binding.
 */
export const keyAliases: Record<string, string> = {
  cmdorctrl: 'mod',
  commandorcontrol: 'mod',
  cmd: 'meta',
  command: 'meta',
  super: 'meta',
  win: 'meta',
  windows: 'meta',
  control: 'ctrl',
  option: 'alt',
  opt: 'alt',
  return: 'enter',
  // A numpad Enter is an Enter. `event.key` already says so; a chord written
  // the long way should reach the same key.
  numpadenter: 'enter',
  esc: 'escape',
  del: 'delete',
  caps: 'capslock'
};

/** One token of a chord, folded: case, spacing and the aliases above. */
export function canonicalKey(token: string): string {
  const normalized = token.toLowerCase().replace(/[\s_-]/g, '');

  return keyAliases[normalized] ?? normalized;
}

/**
 * The handful of keys whose DOM `key` is not what is printed on them.
 *
 * A cap says `↑` and `event.key` says `ArrowUp`; a cap says `Space` and
 * `event.key` says a literal space. Everything else — `Enter`, `Escape`, `Tab`,
 * a letter — already matches once both sides are lower-cased.
 */
const domKeys: Record<string, string> = {
  up: 'arrowup',
  down: 'arrowdown',
  left: 'arrowleft',
  right: 'arrowright',
  space: ' ',
  plus: '+'
};

/**
 * Only the parts of an event a chord is decided on, so the same matcher takes a
 * React synthetic event and a DOM one without either being converted.
 */
export interface PlassKeyState {
  key: string;
  metaKey: boolean;
  ctrlKey: boolean;
  shiftKey: boolean;
  altKey: boolean;
}

/**
 * Does the chord carry a modifier a reader cannot type?
 *
 * `Mod`, `Ctrl`, `Meta` and `Alt` cannot appear in a text field's value, so a
 * chord holding one of them is safe to answer wherever the focus happens to be.
 * **Shift deliberately does not count**: `Shift+A` is how a capital A is typed.
 *
 * This is what lets a global binding be answered while somebody is writing an
 * email without also stealing the letters out of it.
 */
export function hasHardModifier(chord: string): boolean {
  return chord
    .split('+')
    .slice(0, -1)
    .map(canonicalKey)
    .some((part) => part === 'mod' || part === 'ctrl' || part === 'meta' || part === 'alt');
}

/**
 * The keys that do something inside a text field.
 *
 * Everything printable — one code unit of `event.key` is one character a reader
 * just tried to enter — plus the ones that edit or move through what is already
 * there. `Escape`, `Tab` and the function keys are deliberately **not** here:
 * none of them changes a field's value, so a global binding on one is not
 * taking anything away from somebody who is writing.
 */
const editingKeys = new Set([
  'enter',
  'backspace',
  'delete',
  'arrowleft',
  'arrowright',
  'arrowup',
  'arrowdown',
  'home',
  'end',
  'pageup',
  'pagedown'
]);

/** Whether this key would have done something in a field. */
export function editsText(key: string): boolean {
  return key.length === 1 || editingKeys.has(key.toLowerCase());
}

/**
 * Whether the event landed somewhere a reader is entering text.
 *
 * A `<select>` counts: its own typeahead is how a reader picks an option, and a
 * bare letter taken out from under it is a control that stops working. An
 * `<input type="checkbox">` does not — nothing is being typed into it.
 */
export function isTypingTarget(target: EventTarget | null): boolean {
  if (!(target instanceof HTMLElement)) {
    return false;
  }

  if (target.isContentEditable) {
    return true;
  }

  if (target.tagName === 'TEXTAREA' || target.tagName === 'SELECT') {
    return true;
  }

  if (target.tagName === 'INPUT') {
    const type = (target as HTMLInputElement).type;

    return ![
      'button',
      'checkbox',
      'color',
      'file',
      'image',
      'radio',
      'range',
      'reset',
      'submit'
    ].includes(type);
  }

  return false;
}

/**
 * Is this the chord?
 *
 * Every modifier is checked in **both** directions, which is the difference
 * between binding a shortcut and binding a key: `'Enter'` must not fire on
 * `Mod+Enter`, or a field that saves on Enter also saves on every chord that
 * happens to end in one.
 */
export function matchesHotKey(event: PlassKeyState, chord: string): boolean {
  const parts = chord.split('+').map(canonicalKey);
  const wanted = new Set(parts.slice(0, -1));
  const last = parts[parts.length - 1];
  const key = domKeys[last] ?? last;
  const mod = detectOS() === 'mac' ? event.metaKey : event.ctrlKey;

  if (wanted.has('mod') !== mod) {
    return false;
  }

  if (wanted.has('shift') !== event.shiftKey || wanted.has('alt') !== event.altKey) {
    return false;
  }

  // Only when `Mod` was not asked for. It *is* one of these two, and checking
  // both would make `Mod+K` require a Ctrl that a Mac never sends.
  if (!wanted.has('mod')) {
    if (wanted.has('ctrl') !== event.ctrlKey || wanted.has('meta') !== event.metaKey) {
      return false;
    }
  }

  return event.key.toLowerCase() === key;
}

/**
 * The `onKeyDown` a control carrying a `hotKeys` map wears.
 *
 * `undefined` when there is nothing to bind and no handler to call, so a field
 * with neither attaches no listener at all.
 *
 * The caller's own `onKeyDown` runs **first** and unconditionally: it is the
 * lower-level hook and it should see every key, including the ones the map is
 * about to claim. If it consumed the event — `preventDefault()` — the map is
 * skipped, which is how a caller overrules one of their own bindings for a
 * particular key without taking the map apart.
 */
export function hotKeyHandler<E extends Element>(
  hotKeys: PlassHotKeys | undefined,
  onKeyDown: React.KeyboardEventHandler<E> | undefined
): React.KeyboardEventHandler<E> | undefined {
  if (!hotKeys && !onKeyDown) {
    return undefined;
  }

  return (event) => {
    onKeyDown?.(event);

    if (!hotKeys || event.defaultPrevented) {
      return;
    }

    for (const chord of Object.keys(hotKeys)) {
      if (matchesHotKey(event, chord)) {
        // Consumed. The page asked for this key, so it does not also reach the
        // form around the field, the dialog around that, or the browser.
        event.preventDefault();
        event.stopPropagation();
        hotKeys[chord]();

        return;
      }
    }
  };
}
